invisible_cities/cities/beersheba.py

```diff
@@ -0,0 +1,330 @@
+                      iterationThr    : float,
+                      sampleWidth     : List[float],
+                      bin_size        : List[float],
+                      diffusion       : Tuple[float]=(1., 1., 0.3),
+                      iterationGauss  : int=0,
+                      energy_type     : str='Ec',
+                      deconv_mode     : str='joint',
+                      n_dim           : int=2,
+                      cut_type        : str='abs',
+                      interMethod     : str=None):
+
+    """
+    Applies Lucy Richardson deconvolution to SiPM response with a
+    given set of PSFs and parameters.
+
+    Parameters
+    ----------
+    psf_fname       : Point-spread function.
+    ecut            : Cut in relative value to the max voxel over the deconvolution output.
+    iterationNumber : Number of Lucy-Richardson iterations
+    iterationGauss  : Number of Lucy-Richardson iterations for gaussian in 'separate mode'
+    iterationThr    : Stopping threshold (difference between iterations).
+    sampleWidth     : Sampling size of the sensors.
+    bin_size        : Size of the interpolated bins.
+    energy_type     : Energy type ('E' or 'Ec', see Esmeralda) used for assignment.
+    deconv_mode     : 'joint' or 'separate', 1 or 2 step deconvolution, see description later.
+    diffusion       : Diffusion coefficients in each dimension for 'separate' mode.
+    n_dim           : Number of dimensions to apply the method (2 or 3).
+    cut_type        : Cut mode to the deconvolution output ('abs' or 'rel') using ecut
+                      'abs': cut on the absolute value of the hits.
+                      'rel': cut on the relative value (to the max) of the hits.
+    interMethod     : Interpolation method.
+
+    Returns
+    ----------
+    apply_deconvolution : Function that takes hits and returns the
+    deconvolved data.
+    """
+
+    psfs          = pd.read_hdf(psf_fname)
+    deconvolution = deconvolve(iterationNumber, iterationThr, sampleWidth[:n_dim], bin_size[:n_dim], interMethod)
+    dimensions    = ['X', 'Y', 'Z'][:n_dim]
+
+    def create_deconvolution_df(hits, deconvE, pos):
+        df = pd.DataFrame(columns=['event', 'npeak', 'X', 'Y', 'Z', 'E'])
+        if   cut_type == 'abs':
+            selDeconv   = deconvE > ecut
+        elif cut_type == 'rel':
+            selDeconv   = deconvE/deconvE.max() > ecut
+        else:
+            raise TypeError(f'Cut type "{cut_type}" unsupported')
+        ene         = deconvE[selDeconv]
+        df['event'] = [hits.event.unique()[0]] * len(ene)
+        df['npeak'] = [hits.npeak.unique()[0]] * len(ene)
+        df['Z']     = [hits.Z    .unique()[0]] * len(ene) if n_dim == 2 else pos[2][selDeconv]
+        df['X']     = pos[0][selDeconv]
+        df['Y']     = pos[1][selDeconv]
+        df['E']     = ene
+
+        return df
+
+    def distribute_energy(df, hdst):
+        df['E'] = df['E'] / df['E'].sum() * hdst[energy_type].sum()
```

Comment:
We have a situation.
This city uses a function that takes as a parameter the name of a column of a `DataFrame`, which is a `str`. That value comes to the city from the config file so it is also a `str`. However, the name of this column is limited to a few values from an `Enum` object so you find yourself doing something slightly weird:

```
class HitEnergy(AutoNameEnumBase):
    E  = auto()
    Ec = auto()
    Ep = auto()


@city
def beersheva(..., energy_type : str):
    e_type = getattr(HitEnergy, energy_type)
    def do_something(..., etype):
        column = dataframe[etype.value]
    do_something([...], e_type)
```

Notice how you take a `str`, get the corresponding `HitEnergy` object only to then access its `str` value again. The question here is whether there is a benefit to use the `Enum` object or not. On the one hand, you get a bit of documentation on what `energy_type` can be. On the other hand, it adds some lines to the code that may be distracting and not be very informative.

@jacg, I would like to know what's your take on this matter, if you have some time.

Reply:
@gonzaponte Thanks for the summary of the code (very useful), but I'm struggling to find `HitEnergy` anywhere in the real beersheba code!

IIUC, dataframes and the config files impose `str` which makes `HitEnergy` inconvenient. I would like to understand where `HitEnergy` is used throughout the code, before forming an opinion.

Approaches I can imagine, off the top of my head:

+ make the dataframe use the enum instead of strings, and translate it on the way in from the config file ... somewhere
+ accept and use a string in the city, but assert that the string matches one of the enum's members

In general, it would be preferable to use the Enum as much as possible (HitEnergy is so much more informative as a type than str, both for the human and the computer), but if this turns out to be convoluted, then it might be reasonable just to stick to strings. It's a compromise, but I'd need to know more details before forming an opinion.

Reply:
> I'm struggling to find `HitEnergy` anywhere in the real beersheba code!

It had already been defined in [evm/event_model.py](https://github.com/nextic/IC/blob/master/invisible_cities/evm/event_model.py#L116) because it's also used for other pieces of code. The only use for beersheba is in this particular case. But this is actually another good point to use `HitEnergy`: it would be fishy to use `str` here but `HitEnergy` somewhere else when they actually mean the same.

> IIUC, dataframes and the config files impose `str` which makes `HitEnergy` inconvenient

I've just tried and dataframes can in principle take enums as columns both for initialization and column indexing. As for config files that's something that is currently true but that doesn't need to be: the functions that handle this could add all symbols to the context in which the config file is evaluated. For this, we would have to refactor the whole repo a little bit, though.

> I would like to understand where `HitEnergy` is used throughout the code, before forming an opinion.

If you mean in beersheva, there is only this case. For other uses, it is mostly used in [reco/paolina_functions.py](https://github.com/nextic/IC/blob/master/invisible_cities/reco/paolina_functions.py) and in [cities/esmeralda.py](https://github.com/nextic/IC/blob/master/invisible_cities/cities/esmeralda.py), which uses the former module.

> make the dataframe use the enum instead of strings, and translate it on the way in from the config file ... somewhere

I think this is a bit forced... We would have all column names to be strings but that one that is a `HitEnergy`.

> accept and use a string in the city, but assert that the string matches one of the enum's members

That was my second preferred option... right behind leaving it as is.